db/src/db.rs

```diff
@@ -416,45 +437,438 @@ impl DB {
+
+                Ok((e, a, value, value_type_tag,
+                    added,
+                    attribute.index,
+                    attribute.value_type == ValueType::Ref,
+                    attribute.fulltext,
+                    attribute.unique_value))
+            }).collect();
+            let block = block?;
+
+            // `params` reference computed values in `block`.
+            let params: Vec<&ToSql> = block.iter().flat_map(|&(ref e, ref a, ref value, ref value_type_tag, added, index_avet, index_vaet, index_fulltext, unique_value)| {
+                // Avoid inner heap allocation.
+                // TODO: extract some finite length iterator to make this less indented!
+                once(e as &ToSql)
+                    .chain(once(a as &ToSql)
+                           .chain(once(value as &ToSql)
+                                  .chain(once(value_type_tag as &ToSql)
+                                         .chain(once(&tx as &ToSql)
+                                                .chain(once(to_bool_ref(added) as &ToSql)
+                                                       .chain(once(to_bool_ref(index_avet) as &ToSql)
+                                                              .chain(once(to_bool_ref(index_vaet) as &ToSql)
+                                                                     .chain(once(to_bool_ref(index_fulltext) as &ToSql)
+                                                                            .chain(once(to_bool_ref(unique_value) as &ToSql))))))))))
+            }).collect();
+
+            // TODO: cache this for selected values of count.
+            let values: String = repeat_values(values_per_statement, count);
+            let s: String = if search_type == SearchType::Exact {
+                format!("INSERT INTO temp.exact_searches (e0, a0, v0, value_type_tag0, tx0, added0, index_avet0, index_vaet0, index_fulltext0, unique_value0) VALUES {}", values)
+            } else {
+                format!("INSERT INTO temp.inexact_searches (e0, a0, v0, value_type_tag0, tx0, added0, index_avet0, index_vaet0, index_fulltext0, unique_value0) VALUES {}", values)
+            };
+
+            // TODO: consider ensuring we inserted the expected number of rows.
+            let mut stmt = conn.prepare_cached(s.as_str())?;
+            stmt.execute(&params)
+                .map(|_c| ())
+                .chain_err(|| "Could not insert non-fts one statements into temporary search table!")
+        }).collect::<Result<Vec<()>>>();
+
+        results.map(|_| ())
+    }
+
+    /// Take search rows and complete `temp.search_results`.
+    ///
+    /// See https://github.com/mozilla/mentat/wiki/Transacting:-entity-to-SQL-translation.
+    fn search(&self, conn: &rusqlite::Connection) -> Result<()> {
+        // First is fast, only one table walk: lookup by exact eav.
+        // Second is slower, but still only one table walk: lookup old value by ea.
+        let s = r#"
```

Comment:
Generally I have found not, since it adds a level of indirection.  But Firefox does this for some modules, including Places IIRC.  I'd like to leave them more or unless inline until they're more concrete, or truly unmanageable.